Load SSL CA certificate

joelnb opened the pull request:
Currently the :ssl_ca_file option seems to be ignored when creating a Connection. This pull request initialises the CA details according to that option. The option can be specified as a file or directory (meaning that possibly the option should be renamed to reflect that).

My text editor also removed a trailing space but I can redo the pull request without that if you wish to include it and that bothers you.

All tests are currently passing locally using https://github.com/klaemo/docker-couchdb

samlown commented:
Thanks for noticing and suggesting a fix for this! Please include some tests so that I can merge.

joelnb commented:
No problem, 2.0.0 looks like it's going to have some awesome features so thanks for the big update!

Happy to write some tests for this functionality but do you know if they can be run on Travis CI? I've not really used it before but I can't see a way to have an SSL-enabled couchdb server so not too sure how to exercise this functionality.

The only way I can currently think to do it is using the docker functionality of Travis and something like this: https://github.com/klaemo/docker-couchdb-ssl. Not sure if you'd be happy with that though as it introduces an external dependency which could potentially break the tests. 

BTW The reason the tests for this pull request are failing on jruby & ruby 1.9.3 is that 3.0.0 of the mime types gem requires ruby >= 2.0 - https://rubygems.org/gems/mime-types/versions/3.0. I assume this also applies to the master branch of couchrest so if you want to maintain that compatibility you might want to specify a more restrictive dependency for that gem.

samlown commented:
I wouldn't worry about testing with a real server, just ensuring that the
options are passed to the underlying library should be sufficient.

I haven't had much time to look at the failing Travis specs. I'm tempted to
force a minimum version of whatever the mime-types gem requires though.

We've been having a few problems with HTTPClient performance on some Linux
systems, especially inside VirtualBox, so I've not pushed the 2.0.0 release
yet. Having said that, we do use it in production.


On 4 January 2016 at 17:16, Joel <notifications@github.com> wrote:

> No problem, 2.0.0 looks like it's going to have some awesome features so
> thanks for the big update!
>
> Happy to write some tests for this functionality but do you know if they
> can be run on Travis CI? I've not really used it before but I can't see a
> way to have an SSL-enabled couchdb server so not too sure how to exercise
> this functionality.
>
> The only way I can currently think to do it is using the docker
> functionality of Travis and something like this:
> https://github.com/klaemo/docker-couchdb-ssl. Not sure if you'd be happy
> with that though as it introduces an external dependency which could
> potentially break the tests.
>
> BTW The reason the tests for this pull request are failing on jruby & ruby
> 1.9.3 is that 3.0.0 of the mime types gem requires ruby >= 2.0 -
> https://rubygems.org/gems/mime-types/versions/3.0. I assume this also
> applies to the master branch of couchrest so if you want to maintain that
> compatibility you might want to specify a more restrictive dependency for
> that gem.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/couchrest/couchrest/pull/115#issuecomment-168721037>.
>



-- 
samlown.com <http://www.samlown.com>, cabify.com <http://www.cabify.com/>,
autofiscal.com <http://www.autofiscal.com>

joelnb commented:
Great, I'll have a go at that before too long then and update here when I'm done. The way I've implemented it the :ssl_ca_file can take a file or a directory (due to how the HTTPClient method works) so not sure if it should be named differently or just made clear in the docs that this is the case. Do you have any thoughts on this?

joelnb commented:
I have added some tests for this functionality. As the options supports initialisation from a file or directory I have added tests for both cases. The one for the file requires writing a temporary file as the ssl_config initialisation will fail if the file does not exist/does not contain a valid certificate. I used Tempfile so the file should always be cleaned up before exit.

joelnb commented:
The ruby 2.0.0 case is failing due to a bundler issue which is fixed by this commit: https://github.com/bundler/bundler/commit/f4481a7e3dbaf7f563e32afbe2cbbb84e3091f16. If the version of bundler used to install is updated then the 2.0.0 tests will start working as expected again. 

Ruby 1.9.3 & jruby cases are also failing due to the previously-mentioned issue with the mime-types gem. I would personally be fine ditching ruby < 2.0.0 support to align with the mime-types gem but not sure how others would feel. The alternative is to mandate < 3 for the version of the mime-types Gem.

samlown commented:
Looks good! Thanks for this @joelnb! I'm going to try and fix the test issues, support 2.3.0, and update. I agree on not supporting ruby <2.0.0.

joelnb commented:
No problem, thanks for merging it in quickly! If you're interested I have a branch with the test cases all working here: https://github.com/joelnb/couchrest/tree/fix-tests 

I just removed the jruby and 1.9.3 tests and install bundler to ensure the latest version (fixes the issue on 2.2.0).
